src/main/java/edu/ucla/library/iiif/auth/services/DatabaseServiceImpl.java

```diff
@@ -0,0 +1,239 @@
+                return Future.succeededFuture(select.iterator().next().getInteger("access_level"));
+            } else {
+                return Future.failedFuture(LOGGER.getMessage(MessageCodes.AUTH_004, aId, ITEMS));
+            }
+        });
+    }
+
+    @Override
+    public Future<Void> setAccessLevel(final String aId, final int aAccessLevel) {
+        return myDbConnectionPool.withConnection(connection -> {
+            return connection.preparedQuery(UPSERT_ACCESS_LEVEL).execute(Tuple.of(aId, aAccessLevel));
+        }).recover(error -> {
+            return Future.failedFuture(LOGGER.getMessage(MessageCodes.AUTH_006, UPSERT_ACCESS_LEVEL, error));
+        }).compose(upsert -> {
+            if (hasSingleRow(upsert)) {
+                if (upsert.iterator().hasNext()) {
+                    LOGGER.debug(upsert.iterator().next().deepToString());
+                }
+                return Future.succeededFuture();
+            } else {
+                return Future.failedFuture(LOGGER.getMessage(MessageCodes.AUTH_005, aId, ITEMS));
+            }
+        });
+    }
+
+    @Override
+    public Future<Boolean> getDegradedAllowed(final String aOrigin) {
+        return myDbConnectionPool.withConnection(connection -> {
+            return connection.preparedQuery(SELECT_DEGRADED_ALLOWED).execute(Tuple.of(aOrigin));
+        }).recover(error -> {
+            return Future.failedFuture(LOGGER.getMessage(MessageCodes.AUTH_006, SELECT_DEGRADED_ALLOWED, error));
+        }).compose(select -> {
+            if (hasSingleRow(select)) {
+                return Future.succeededFuture(select.iterator().next().getBoolean("degraded_allowed"));
+            } else {
+                return Future.failedFuture(LOGGER.getMessage(MessageCodes.AUTH_004, aOrigin, ORIGINS));
+            }
+        });
+    }
+
+    @Override
+    public Future<Void> setDegradedAllowed(final String aOrigin, final boolean aDegradedAllowed) {
+        return myDbConnectionPool.withConnection(connection -> {
+            return connection.preparedQuery(UPSERT_DEGRADED_ALLOWED).execute(Tuple.of(aOrigin, aDegradedAllowed));
+        }).recover(error -> {
+            return Future.failedFuture(LOGGER.getMessage(MessageCodes.AUTH_006, UPSERT_DEGRADED_ALLOWED, error));
+        }).compose(upsert -> {
+            if (hasSingleRow(upsert)) {
+                if (upsert.iterator().hasNext()) {
+                    LOGGER.debug(upsert.iterator().next().deepToString());
+                }
+                return Future.succeededFuture();
+            } else {
+                return Future.failedFuture(LOGGER.getMessage(MessageCodes.AUTH_005, aOrigin, ORIGINS));
+            }
+        });
+    }
+
+    @Override
+    public Future<PgPool> getConnectionPool() {
+        return Future.succeededFuture(myDbConnectionPool);
+    }
+
+    @Override
+    public Future<Redis> getRedisClient() {
+        return Future.succeededFuture(myDbCacheClient);
+    }
```

Comment:
Apparently the service proxy pattern requires all the interface methods to either (1) return a Future, or (2) return void and accept a callback as a parameter. That choice is made via `useFutures` in the package-info.java.

Reply:
Yes, this is true. I asked about this on the Vert.x discord not too long ago.